packages/core/contracts/bot-helpers/uniswap-broker/UniswapBrokerV3.sol

```diff
@@ -0,0 +1,631 @@
+            } else if (!zeroForOne) {
+                step.sqrtPriceNextX96 = nextTickPriceX96 > sqrtRatioTargetX96 ? nextTickPriceX96 : sqrtRatioTargetX96;
+                inputAmountForStep = SqrtPriceMath.getAmount1Delta( // As we are trading token1 for token0, calculate the token1 input.
+                    step.sqrtPriceStartX96,
+                    step.sqrtPriceNextX96,
+                    state.liquidity,
+                    false
+                );
+            }
+
+            // Add amount for this step to the total required input.
+            state.requiredInputAmount = state.requiredInputAmount.add(inputAmountForStep);
+
+            // If we have hit(or exceeded) our target price in the associate direction, then stop.
+            if (zeroForOne && state.sqrtPriceX96 <= sqrtRatioTargetX96) break;
+            if (!zeroForOne && state.sqrtPriceX96 >= sqrtRatioTargetX96) break;
+
+            // If the next step is is initialized then we will need to update the liquidity for the current step.
+            if (step.initialized) {
+                // Fetch the net liquidity. this could be positive or negative depending on if a LP is turning on or off at this price.
+                (, int128 liquidityNet, , , , , , ) = pool.ticks(step.tickNext);
+                if (!zeroForOne) liquidityNet = -liquidityNet;
+
+                state.liquidity = LiquidityMath.addDelta(state.liquidity, liquidityNet);
+            }
+
+            // Finally, set the state price to the next price for the next iteration.
+            state.sqrtPriceX96 = step.sqrtPriceNextX96;
+            state.tick = step.tickNext;
+        }
+
+        // Based on the direction we are moving, set the input and output tokens.
+        address tokenIn = zeroForOne ? pool.token0() : pool.token1();
+        address tokenOut = zeroForOne ? pool.token1() : pool.token0();
+
+        // If trading from an EOA pull tokens into this contract. If trading from a DSProxy this is redundant.
+        if (tradingAsEOA)
+            TransferHelper.safeTransferFrom(tokenIn, msg.sender, address(this), state.requiredInputAmount);
+
+        // Approve the router and execute the swap.
+        TransferHelper.safeApprove(tokenIn, address(uniswapRouter), state.requiredInputAmount);
+        ISwapRouter(uniswapRouter).exactInputSingle(
+            ISwapRouter.ExactInputSingleParams({
+                tokenIn: tokenIn,
+                tokenOut: tokenOut,
+                fee: pool.fee(),
+                recipient: recipient,
+                deadline: deadline,
+                amountIn: state.requiredInputAmount,
+                amountOutMinimum: 0,
+                sqrtPriceLimitX96: sqrtRatioTargetX96
+            })
+        );
+
+        return state.requiredInputAmount;
+    }
+}
+
+// The two methods below are taken almost verbatim from https://github.com/Uniswap/uniswap-v3-core/blob/main/contracts/libraries/TickBitmap.sol.
```

Comment:
the rest of this smart contract is libraries and helpers taken from Uniswap and updated to accommodate solidity 0.8.0. Ideally, we would not need to have these in this file and could just import them. However, as we currently still use truffle we need _all_ contracts to be on the same version. _some_ of the solidity files from uniswap are bound to `0.7.x` versions and these are incompatible with our repo. 

once we get rid of truffle completely we will be able to ditch all this code and just import the libs :)